include/eve/detail/function/simd/riscv/slice.hpp

```diff
@@ -0,0 +1,187 @@
+      else if constexpr( out_lmul == -2 ) return __riscv_vlmul_trunc_i8mf2(a);
+      else if constexpr( out_lmul == 1 ) return __riscv_vlmul_trunc_i8m1(a);
+      else if constexpr( out_lmul == 2 ) return __riscv_vlmul_trunc_i8m2(a);
+      else if constexpr( out_lmul == 4 ) return __riscv_vlmul_trunc_i8m4(a);
+    }
+    else if constexpr( match(c, category::uint8) )
+    {
+      if constexpr( out_lmul == -8 ) return __riscv_vlmul_trunc_u8mf8(a);
+      else if constexpr( out_lmul == -4 ) return __riscv_vlmul_trunc_u8mf4(a);
+      else if constexpr( out_lmul == -2 ) return __riscv_vlmul_trunc_u8mf2(a);
+      else if constexpr( out_lmul == 1 ) return __riscv_vlmul_trunc_u8m1(a);
+      else if constexpr( out_lmul == 2 ) return __riscv_vlmul_trunc_u8m2(a);
+      else if constexpr( out_lmul == 4 ) return __riscv_vlmul_trunc_u8m4(a);
+    }
+  }
+}
+
+//================================================================================================
+// Single slice
+//================================================================================================
+
+template<callable_options O, typename T, typename N>
+EVE_FORCEINLINE wide<T, typename N::split_type>
+                slice_(EVE_REQUIRES(rvv_), O const&, wide<T, N> a, lower_slice_t) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  constexpr auto in_lmul  = rvv_lmul_v<T, N>;
+  constexpr auto out_lmul = rvv_lmul_v<T, typename N::split_type>;
+  if constexpr( in_lmul == out_lmul ) return a.storage();
+  else
+  {
+    // we need to lower lmul - call lmul trunc.
+    return riscv_lmul_trunc(a);
+  }
+}
+
+template<callable_options O, typename T, typename N>
+EVE_FORCEINLINE wide<T, typename N::split_type>
+                slice_(EVE_REQUIRES(rvv_), O const&, wide<T, N> a, upper_slice_t) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  constexpr auto shift_size = N::split_type::value;
+  wide<T, N>     res        = __riscv_vslidedown(a, shift_size, N::value);
+  constexpr auto in_lmul    = rvv_lmul_v<T, N>;
+  constexpr auto out_lmul   = rvv_lmul_v<T, typename N::split_type>;
+  if constexpr( in_lmul == out_lmul ) return res.storage();
+  else
+  {
+    // we need to lower lmul - call lmul trunc.
+    return riscv_lmul_trunc(res);
+  }
+}
+
+template<callable_options O, typename T, typename N>
+EVE_FORCEINLINE logical<wide<T, typename N::split_type>>
+                slice_(EVE_REQUIRES(rvv_), O const&, logical<wide<T, N>> a, lower_slice_t) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  return simd_cast(a, as<logical<wide<T, typename N::split_type>>> {});
+}
+
+template<callable_options O, typename T, typename N>
+EVE_FORCEINLINE logical<wide<T, typename N::split_type>>
+                slice_(EVE_REQUIRES(rvv_), O const&, logical<wide<T, N>> a, upper_slice_t) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  auto bits_slice = a.bits().slice(upper_slice_t {});
+
+  auto neq = bits_slice != 0;
+  return simd_cast(neq, as<logical<wide<T, typename N::split_type>>> {});
```

Comment:
seems like you don't need the `!= 0`

If you simd_cast to m1 uints, slice and simd_cast back. Instead of doing `bits()`

Reply:
> If you simd_cast to m1 uints

This method will not work, if you have `N<=8` (So you have some undefined bits even for `char`).

>  simd_cast 

For now I forbid for `simd_cast` casting between `logical` and `wide`. I think this could be added when I will create an MR with optimization for logical insert/extract (for now, as discussed, we will use memory to do this)

Reply:
So - if `n < 8` - do a bit shift, if `n >= 8` -> cast and slice?
I'm ok if you don't want to do the logical simd cast now - put a todo here please.

Reply:
Added TODO for now